File: Students/John/python/lab08.py
"""

all codes are written and created by John Robson Wed Mar 10, 2021

"""

import random

# Version 1

def runVersion1():
    x = random.randint(1, 10)
    count = 1

    while True:
        if count == 11:
            print("You lose")
            break
        guess = int(input("Choose a number:"))

        if (guess == x):
            print("You win!")
        else:
            print(f"Try again. You have guessed {count} time(s).")

        count += 1


# Version 2

def runVersion2():
    x = random.randint(1, 10)
    count = 1

    while True:
        guess = int(input("Choose a number?"))

        if (guess == x):
            print("You win!")
            break
        else:
            print(f"Try again. You have guessed {count} times.")

        count += 1


# Version 3

def runVersion3():
    x = random.randint(1, 10)
    count = 1

    while True:
        guess = int(input("Choose a number?"))

        if (guess == x):
            print("You win!")
            break
        else:
            if guess > x:
                print("Too high!")
            else:
                print("Too low!")
            print(f"Try again. You have guessed {count} times.")

        count += 1


# Version 4

def runVersion4():
    x = random.randint(1, 10)
    guess = -1
    count = 1
    lastGuess = 0

    while True:
        guess = int(input("Choose a number?"))

        if (guess == x):
            print("You win!")
            break
        else:
            if lastGuess != 0:
                if abs(x - lastGuess) < abs(x - guess):
                    print("Last guess was closer than current")
                else:
                    print("Current guess is closer than last")
            if guess > x:
                print("Too high!")
            else:
                print("Too low!")
            print(f"Try again. You have guessed {count} times.")

        lastGuess = guess
        count += 1


# Version 5

def runVersion5():
    x = int(input("Pick a number 1-10"))
    count = 1

    if x < 1 or x > 10:
        print("Your number was out of the range. Try again:")
        runVersion5()
    else:
        while True:
            guess = random.randint(1, 10)

            if (guess == x):
                print("You win!")
                break
            else:
                print(f"Try again. You have guessed {count} times.")

            count += 1


versions = [runVersion1, runVersion2, runVersion3, runVersion4, runVersion5]

while True:
    strategy = 0

    print('Welcome. You can type "done" at any time to exit.')

    while strategy < 1 or strategy > 5:
        strategy = int(input("Which version would you like to run: 1 - 5"))
        if strategy == "done":
            break
        
    if strategy == "done":
        break

    strategy = versions[strategy - 1]

    strategy()

File: Students/John/javascript/lab08.js
/*

all codes are written and created by John Robson Fri Mar 19, 2021

*/

let guessCount = 0,
  over = "false";

export function runLab08() {
  renApp(
    [
      { el: "h1", text: "Lab 08 - Guess The Number" },
      { el: "label", text: "Guess a number 1 - 9" },
      { el: "input", cl: "guess", type: "number", min: 1, max: 9 },
      { el: "input", cl: "comp-guess", type: "hidden" },
      { el: "h2", cl: "grade" },
      { el: "aside", cl: "history" },
      { el: "button", cl: "reset", text: "reset" },
    ],
    $("main")
  );

  renApp(
    [
      { el: "h2", text: "History" },
      { el: "div", cl: "hist-container" },
    ],
    $(".history")
  );

  $(".guess").addEventListener("input", () => {
    checkInput($(".guess"), 1);
  });
  $(".guess").onchange = checkGuess;

  $(".comp-guess").value = randN(1, 9);
  $(".reset").onclick = reset;
}

const checkGuess = () => {
  checkValue($(".guess"), 1, 9);
  if ($(".guess").value == null || $(".guess").value === "") return;

  if (guessCount >= 5) {
    gameOver();
    return;
  }

  let status = "";

  if ($(".guess").value === $(".comp-guess").value) status = "You win";
  else
    status =
      $(".guess").value > $(".comp-guess").value ? "Too high" : "Too low";

  guessCount++;
  addToHistory(status);
  if (status === "You win") over = "true";
  if (guessCount >= 5) {
    gameOver();
    return;
  }
};

const reset = () => {
  checkValue($(".guess"), 1, 9);
  clear($(".hist-container"));
  $(".guess").value = "";
  guessCount = 0;
  over = "false";
};

const addToHistory = (status) => {
  if (over === "true") return;
  limit($$(".hist-item"), 5);

  $(".grade").innerText.split(": ")[1];

  let query = `${$(".guess").value}: ${status}`;
  renApp({ el: "p", cl: "hist-item", text: query }, $(".hist-container"));
};

const gameOver = () => {
  if (over !== "true") {
    over = "true";
    renApp(
      {
        el: "p",
        cl: "hist-item",
        text: "You Lose! Answer: " + $(".comp-guess").value,
      },
      $(".hist-container")
    );
  }
};
